internal/widget/base.go

```diff
@@ -10,34 +12,72 @@ type base struct {
 	hidden bool
 	pos    fyne.Position
 	size   fyne.Size
+
+	propertyLock sync.RWMutex
 }
 
 // Move satisfies the fyne.Widget interface.
 func (b *base) Move(pos fyne.Position) {
-	b.pos = pos
+	b.setFieldsAndRefresh(nil, func() {
+		b.pos = pos
+	})
 }
 
 // Position satisfies the fyne.Widget interface.
 func (b *base) Position() fyne.Position {
-	return b.pos
+	var pos fyne.Position
+	b.readFields(func() {
+		pos = b.pos
+	})
+	return pos
+}
+
+// readFields provides a guaranteed thread safe way to access widget fields.
+func (b *base) readFields(f func()) {
+	b.propertyLock.RLock()
+	defer b.propertyLock.RUnlock()
+
+	f()
+}
+
+// SetFieldsAndRefresh helps to make changes to a widget that should be followed by a refresh.
+// This method is a guaranteed thread-safe way of directly manipulating widget fields.
+func (b *base) setFieldsAndRefresh(w fyne.Widget, f func()) {
+	b.propertyLock.Lock()
+	f()
+	b.propertyLock.Unlock()
+
+	if w != nil { // the wrapping function didn't tell us what to refresh
+		b.refresh(w)
+	}
 }
 
 // Size satisfies the fyne.Widget interface.
 func (b *base) Size() fyne.Size {
-	return b.size
+	var size fyne.Size
+	b.readFields(func() {
+		size = b.size
+	})
+	return size
 }
 
 // Visible satisfies the fyne.Widget interface.
 func (b *base) Visible() bool {
-	return !b.hidden
+	var hidden bool
+	b.readFields(func() {
+		hidden = b.hidden
+	})
+	return !hidden
 }
 
 func (b *base) hide(w fyne.Widget) {
-	if b.hidden {
+	if !b.Visible() {
 		return
 	}
 
+	b.propertyLock.Lock()
 	b.hidden = true
+	b.propertyLock.Unlock()
```

Comment:
Use setFieldsAndRefresh(f func(){...}, nil) for consistency

Reply:
This is diferent. We don't actually refresh widget on hide - just the canvas refresh expressed below.

Reply:
Hence the nil - you already have a nil check in setFieldsAndRefresh

Reply:
Oh i missed that. The nil check was really safety rather than a feature I thought should be added. using a "nil" special value is generally an anti-pattern so I would prefer an additional method over telling people to pass nil